drivers: adc: fix functionality gaps

pabigot opened the pull request:
Two commits:
* Provide API to retrieve the internal reference voltage, which is required to make sense of results that use that reference source.  Support is implemented in Nordic ADC and SAADC; see #11922 for other platforms that might need this functionality.
* Provide API to specify that the ADC should be calibrated before performing the sample.  This functionality is implemented on Nordic SAADC, including workarounds for relevant PANs.

Closes #11922

pabigot committed:
drivers: adc: add API to retrieve internal reference voltage

Without this there's no point providing ADC_REF_INTERNAL as an option.

Relates-to: issue #11922
Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: adc: add API to support calibration

Add a flag to the sequence structure that tells the driver it should
calibrate the ADC prior to initiating the sample.

Implement this for nRF SAADC.  The implementation supports the
workarounds for PAN-86 and PAN-178.

Relates-to: issue #11922
Signed-off-by: Peter A. Bigot <pab@pabigot.com>

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -358,6 +360,20 @@ static void saadc_irq_handler(void *param)
+		nrf_saadc_event_clear(NRF_SAADC_EVENT_CALIBRATEDONE);
+		nrf_saadc_event_clear(NRF_SAADC_EVENT_STOPPED);
+		nrf_saadc_task_trigger(NRF_SAADC_TASK_STOP);
+		while (!nrf_saadc_event_check(NRF_SAADC_EVENT_STOPPED)) {
```
I don't entirely understand the purpose of this part of code. Why can't you just start the sampling after getting (and clearing) the CALIBRATEDONE event?
Such loop in the interrupt handler does not look nice.

anangl replied:
Oh, I see it now, PAN-86. A word of comment here could help.

anangl commented on include/adc.h:
```diff
@@ -294,6 +319,26 @@ struct adc_driver_api {
+/**
+ * @brief Retrieve the internal reference voltage
+ *
+ * @param dev Pinter to the device structure for the driver instance.
```
Pointer

anangl commented on include/adc.h:
```diff
@@ -294,6 +319,26 @@ struct adc_driver_api {
+ * @param dev Pinter to the device structure for the driver instance.
+ *
+ * @return The internal reference, in millivolts.
+ * @retval -ENOTSUP if there is no internal reference
```
I guess `@return` and `@retval` should not be used together. Wouldn't it be better to just add ", or -ENOTUP if ..." to the description above?

pabigot replied:
It would be better if `@return` and `@retval` could be used together, so the explanation of how each error value should be interpreted could be provided.  However, we must deal with the world as it is, not as we wish it would be....

Done.

anangl commented on include/adc.h:
```diff
@@ -294,6 +319,26 @@ struct adc_driver_api {
+	int rv = -ENOTSUP;
+
+	const struct adc_config_info *cip
+		= (struct adc_config_info *)dev->config->config_info;
```
`const` missing

pabigot replied:
And this is why I prefer C++, which wouldn't have let that get past the compiler.

Done, as with all other requests.  Thanks for the review; let me know if you see anything remaining in the update.

pabigot commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -374,6 +390,8 @@ static int init_saadc(struct device *dev)
 	nrf_saadc_event_clear(NRF_SAADC_EVENT_END);
 	nrf_saadc_int_enable(NRF_SAADC_INT_END);
+	nrf_saadc_event_clear(NRF_SAADC_EVENT_CALIBRATEDONE);
+	nrf_saadc_int_enable(NRF_SAADC_INT_CALIBRATEDONE);
```
This is why I don't like nrfx hal: if I could just do:
```
NRF_SAADC->INTEN = (SAADC_INTENSET_CALIBRATEDONE_Set << SAADC_INTENSET_CALIBRATEDONE_Pos)
                 | (SAADC_INTENSET_END_Set << SAADC_INTENSET_END_Pos);
```
I'd know what was going to happen, rather than having to dig into the HAL source to see what `nrfc_saadc_int_enable` does with its arguments.

Done.

Sizurka commented on include/adc.h:
```diff
@@ -294,6 +309,26 @@ struct adc_driver_api {
+	int rv = -ENOTSUP;
+
+	const struct adc_config_info *cip
+		= (const struct adc_config_info *)dev->config->config_info;
```
Doesn't the `struct adc_config_info` need to be added to the other drivers that currently use `config_info` for their instance specific information, even if it's just populated with `-ENOTSUP`.  That is, wouldn't this currently return nonsense for them? (looks like mostly it would end up returning the base address of the instance registers from the current struct layouts).  Granted, it's a new API so nothing using them would currently fail, but it would be an odd result if something new tried to use it.

pabigot commented:
I modeled this on the counter API which has similar constants that are device specific, but as @Sizurka points out this doesn't work for ADC because only Nordic devices left this field unused.  Other drivers are using `config_info` for internal data structures.  I'm not going to mess with those, so this'll have to be solved another way.

pabigot commented:
I've removed the reference voltage API that's incompatible with other drivers.

I also confirmed that `EVENTS_STOPPED` occurs less than 200 ns after `TASKS_STOP` is triggered so I've removed the loop waiting for that condition.  This is consistent with how start and stop is handled in the existing driver, so should be less confusing.
